[17.0][MIG] account_billing: Migration to 17.0

AungKoKoLin1997 opened the pull request:
Supersedes https://github.com/OCA/account-invoicing/pull/1763

@qrtl QT5088

Saran440 committed:
[12.0][ADD] account_billing

oca-travis committed:
[UPD] Update account_billing.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

kongrattapong committed:
[IMP] account_billing: black, isort

kongrattapong committed:
[MIG] account_billing: Migration to 13.0

komsan-S committed:
[IMP] : black, isort, prettier

komsan-S committed:
[MIG] account_billing: Migration to 14.0

oca-travis committed:
[UPD] Update account_billing.pot

OCA-git-bot committed:
[UPD] README.rst

anasuarez1 committed:
Added translation using Weblate (Spanish)

JordiBForgeFlow committed:
[14.0][fix][account_billing] fixes failing tests

Yvesldff committed:
Added translation using Weblate (French (France))

OCA-git-bot committed:
account_billing 14.0.1.0.1

joao-p-marques committed:
[FIX] account_billing: Fix tests

After https://github.com/odoo/odoo/commit/fcaa54939e9a4f0dd5e47cd0ccffe7aa24bd451c, we need to set the invoice date manually.

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 0.9% (1 of 105 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_billing/zh_CN/

mbcosta committed:
[FIX][BUG] Get the current company should be made with self.env.company, module account_billing.

OCA-git-bot committed:
account_billing 14.0.1.0.2

OCA-git-bot committed:
[IMP] update dotfiles

MiquelRForgeFlow committed:
[FIX] account_billing: minor fixes

oca-travis committed:
[UPD] Update account_billing.pot

OCA-git-bot committed:
account_billing 14.0.1.0.3

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-14.0/account-invoicing-14.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_billing/

bizzappdev committed:
[MIG][15.0]account_billing:Migration to 15.0

bosd committed:
Added translation using Weblate (Dutch (nl_NL))

bosd committed:
Translated using Weblate (Dutch (nl_NL))

Currently translated at 90.4% (95 of 105 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_billing/nl_NL/

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-14.0/account-invoicing-14.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_billing/

Commit:
[UPD] Update account_billing.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-14.0/account-invoicing-14.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_billing/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (104 of 104 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_billing/es/

Saran440 committed:
[IMP] account_billing: pre-commit stuff

Saran440 committed:
[MIG] account_billing: Migration to 16.0

Saran440 committed:
[IMP] account_billing: add feature register payment in billing and change UX billing

Commit:
[UPD] Update account_billing.pot

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (107 of 107 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_billing/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (107 of 107 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_billing/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (107 of 107 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_billing/it/

AungKoKoLin1997 committed:
[IMP] account_billing: Add multi-company record rule

AungKoKoLin1997 committed:
[IMP] account_billing: Add billing_ids to invoice list

Co-authored-by: Yoshi Tashiro <tashiro@quartile.co>

Ryoko04 committed:
Added translation using Weblate (Japanese)

Ryoko04 committed:
Translated using Weblate (Japanese)

Currently translated at 30.8% (33 of 107 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_billing/ja/

Ryoko04 committed:
Translated using Weblate (Japanese)

Currently translated at 89.7% (96 of 107 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_billing/ja/

Ryoko04 committed:
Translated using Weblate (Japanese)

Currently translated at 99.0% (106 of 107 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_billing/ja/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (107 of 107 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_billing
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_billing/it/

AungKoKoLin1997 committed:
[FIX] account_billing: fix domain of _get_move()

santostelmo committed:
[Fix] account_billing test_create_billing_from_selected_invoices

AungKoKoLin1997 committed:
[FIX] account_billing: use single record instead of recordset

AungKoKoLin1997 committed:
[IMP] account_billing: remove threshold_date field from billing line and invoice_date as compute field

AungKoKoLin1997 committed:
[IMP] account_billing: not allow rest to draft for move that is related with  non cancelled billing

OCA-git-bot committed:
[BOT] post-merge updates

AungKoKoLin1997 committed:
[IMP] account_billings: remove development_status

AungKoKoLin1997 committed:
[IMP] account_billing: pre-commit stuff

AungKoKoLin1997 committed:
[MIG] account_billing: Migration to 17.0

yostashiro commented on account_billing/models/account_billing.py:
```diff
@@ -0,0 +1,255 @@
+    def action_cancel(self):
+        for rec in self:
+            invoice_paid = rec.billing_line_ids.mapped("move_id").filtered(
+                lambda m: m.payment_state == "paid"
```
Any reason to change the param to `m`?

AungKoKoLin1997 replied:
Yes. There is pre-commit issue if we use existing variable `l`.

yostashiro commented on account_billing/tests/test_account_billing.py:
```diff
@@ -0,0 +1,270 @@
+from odoo.tests.common import TransactionCase
+
+
+class TestAccountBilling(TransactionCase):
```
Revive `@tagged("post_install", "-at_install")`.

yostashiro commented on account_billing/tests/test_account_billing.py:
```diff
@@ -0,0 +1,270 @@
+        cls.billing_model = cls.env["account.billing"]
+        cls.register_payments_model = cls.env["account.payment.register"]
+
+        cls.partner_id = cls.env["res.partner"].create({"name": "Test Partner"})
```
```suggestion
        cls.partner = cls.env["res.partner"].create({"name": "Test Partner"})
```

yostashiro commented on account_billing/tests/test_account_billing.py:
```diff
@@ -0,0 +1,270 @@
+        cls.invoice_line_model = cls.env["account.move.line"]
+        cls.billing_model = cls.env["account.billing"]
+        cls.register_payments_model = cls.env["account.payment.register"]
+
```
```suggestion
```

yostashiro commented on account_billing/tests/test_account_billing.py:
```diff
@@ -0,0 +1,270 @@
+        cls.currency_eur = cls.env.ref("base.EUR")
+        cls.currency_eur.active = True
+        cls.currency_usd_id = cls.env.ref("base.USD").id
+        cls.currency_eur_id = cls.currency_eur.id
```
Just to be consistent and to avoid having unnecessary objects.
```suggestion
        cls.currency_usd = cls.env.ref("base.USD")
        cls.currency_usd.active = True
```

yostashiro commented on account_billing/tests/test_account_billing.py:
```diff
@@ -0,0 +1,270 @@
+                "partner_id": partner or self.partner_agrolait.id,
+                "currency_id": currency_id or self.currency_eur_id,
+                "move_type": invoice_type,
+                "invoice_date": fields.Date.context_today(self.env.user),
```
I think it's better to have `invoice_date` as a method param to allow some flexibility in test scenarios.

AungKoKoLin1997 committed:
[IMP] account_billing: imp tests

yostashiro reviewed (approved):
Code review and functional test. LGTM.

AungKoKoLin1997 committed:
[FIX] account_billing: fix wrong currency assigned in billing

This commit fixes the issue where the billing incorrectly used the current company’s currency instead of the moves’ currency
when creating billings from the account move list view for moves in a different currency.

BhaveshHeliconia commented:
@AungKoKoLin1997 : I think the commit history is not proper. After the migration commit, I see separate fix and improvement commits, which should be squashed into the migration commit. The last commit should contain the complete migration.

AungKoKoLin1997 commented:
@BhaveshHeliconia The commits after the migration commit are forward ports from unmerged PR (https://github.com/OCA/account-invoicing/pull/1986), along with some new commits that will later be backported or forward ported.
That’s why I think keeping the commits split is better for backporting and keeping the same commit history.
For this reason, I didn’t include them in the migration commit and kept the commit history as it is.

BhaveshHeliconia commented:
@AungKoKoLin1997 : OK Great! Thanks for your contribution.

BhaveshHeliconia reviewed (approved):
Functional test LGTM!

AungKoKoLin1997 committed:
[FIX] account_billing: fix issue when create billings from refunds or receipts

yostashiro commented:
@AungKoKoLin1997 Please add https://github.com/OCA/account-invoicing/commit/6c8ebef5f24b5d95d4b3b9202082c6c92330e55a

yostashiro commented on account_billing/views/account_billing_views.xml:
```diff
@@ -0,0 +1,300 @@
+                            />
+                            <field name="billing_line_ids" readonly="state != 'draft'">
+                                <tree name="billing" create="0" editable="bottom">
+                                    <field name="move_id" invisible="1" />
```
The same thing we fixed for 18.0.
```suggestion
                                    <field name="move_id" column_invisible="1" />
```

yostashiro commented on account_billing/views/account_billing_views.xml:
```diff
@@ -0,0 +1,300 @@
+                                    <field name="name" />
+                                    <field name="invoice_date" />
+                                    <field name="origin" />
+                                    <field name="currency_id" invisible="1" />
```
```suggestion
                                    <field name="currency_id" column_invisible="1" />
```

yostashiro reviewed (changes requested):
The PR currently includes an unrelated module. Please remove it.

AungKoKoLin1997 committed:
Add Aung Ko Ko Lin as module maintainer

AungKoKoLin1997 commented:
I added myself as a maintainer of this module as there are multiple PRs related to it, and I need to follow up on forward ports and ongoing changes.

yostashiro reviewed (approved):
Code review. LGTM.

smorita7749 committed:
[IMP] account_billing: always show invoice relate billing stat button

smorita7749 commented:
Add a commit to always display invoice relate billing stat button.

SanamiK commented:
Functional review. LGTM.

smorita7749 reviewed (approved):
Functional review: LGTM
Billing creation via the compute items button in billing form and the create billing button in action button is properly working.
